File: opcua-webservice-spring/core/src/main/java/org/opcfoundation/webserver/types/common/UaBrowseContinuationPoint.java
package org.opcfoundation.webserver.types.common;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.eclipse.milo.opcua.stack.core.types.builtin.ByteString;
import org.eclipse.milo.opcua.stack.core.types.builtin.NodeId;
import org.eclipse.milo.opcua.stack.core.types.builtin.unsigned.UInteger;
import org.eclipse.milo.opcua.stack.core.types.enumerated.BrowseDirection;
import org.eclipse.milo.opcua.stack.core.types.structured.BrowseDescription;
import org.springframework.lang.Nullable;

public class UaBrowseContinuationPoint {
    private static final ObjectMapper jsonMapper = new ObjectMapper();

    private String id;
    private Integer bDir;
    private String refId;
    private Boolean includeSt;
    private Integer ncMask;
    private Integer rltMask;
    private Integer maxRef;
    private Integer offset;
    private Integer taskMsk;

    public UaBrowseContinuationPoint()
    {
        id = "";
        bDir = 0;
        refId = "";
        includeSt = false;
        ncMask = 0;
        rltMask = 0;
        maxRef = 0;
        offset = 0;
        taskMsk = 0;
    }

    public UaBrowseContinuationPoint(
            BrowseDescription description,
            UaBrowseAdditionalInfo additionalInfo)
    {
        id = description.getNodeId().toParseableString();
        bDir = description.getBrowseDirection().getValue();
        refId = description.getReferenceTypeId().toParseableString();
        includeSt = description.getIncludeSubtypes();
        ncMask = description.getNodeClassMask().intValue();
        rltMask = description.getResultMask().intValue();
        maxRef = additionalInfo.getMaxReferencesPerNode();
        offset = additionalInfo.getReferenceOffset();
        taskMsk = additionalInfo.getTaskCheckListMasks();
    }

    public @Nullable BrowseDescription browseDescription()
    {
        NodeId nodeId = NodeId.parseOrNull(id);
        BrowseDirection browseDirection = BrowseDirection.from(bDir);
        NodeId referenceTypeId = NodeId.parseOrNull(refId);
        if (null == nodeId || null == browseDirection || 0 > ncMask || 0 > rltMask) return null;
        return new BrowseDescription(nodeId,browseDirection, referenceTypeId, includeSt, UInteger.valueOf(ncMask), UInteger.valueOf(rltMask));
    }

    public UaBrowseAdditionalInfo additionalInfo()
    {
        return new UaBrowseAdditionalInfo(maxRef, offset, taskMsk);
    }

    public ByteString toByteString()
    {
        try
        {
            return ByteString.of(jsonMapper.writeValueAsBytes(this));
        } catch (Exception e) {
            return ByteString.NULL_VALUE;
        }
    }

    public static @Nullable UaBrowseContinuationPoint fromByteString(ByteString continuationPoint)
    {
        try {
            if (continuationPoint.isNull()) return null;
            return jsonMapper.readValue(continuationPoint.bytes(), UaBrowseContinuationPoint.class);
        } catch (Exception e) {
            return null;
        }
    }

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public Integer getbDir() {
        return bDir;
    }

    public void setbDir(Integer bDir) {
        this.bDir = bDir;
    }

    public String getRefId() {
        return refId;
    }

    public void setRefId(String refId) {
        this.refId = refId;
    }

    public Boolean getIncludeSt() {
        return includeSt;
    }

    public void setIncludeSt(Boolean includeSt) {
        this.includeSt = includeSt;
    }

    public Integer getRltMask() {
        return rltMask;
    }

    public void setRltMask(Integer rltMask) {
        this.rltMask = rltMask;
    }

    public Integer getNcMask() {
        return ncMask;
    }

    public void setNcMask(Integer ncMask) {
        this.ncMask = ncMask;
    }

    public Integer getMaxRef() {
        return Math.max(maxRef,0);
    }

    public void setMaxRef(Integer maxRef) {
        this.maxRef = maxRef;
    }

    public Integer getOffset() { return offset; }

    public void setOffset(Integer offset) {
        this.offset = offset;
    }

    public Integer getTaskMsk() { return taskMsk; }

    public void setTaskMsk(Integer taskMsk) { this.taskMsk = taskMsk; }
}

File: opcua-webservice-node/source/src/server/types/common/UaBrowseContinuationPoint.ts
import { BrowseDirection } from "opcua-webapi";
import { parseUaNodeIdOrNull, UaBrowseDescription, UaNodeId } from "opcua-webapi-ts";
import { UaBrowseAdditionalInfo } from "./UaBrowseAdditionalInfo";

export type SerializedBrowseContinuationPoint = {
    id: string;
    bDir: number;
    refId: string;
    includeSt: boolean;
    ncMask: number;
    rltMask: number;
    maxRef: number;
    offset: number;
    taskMsk: number;
};

export class UaBrowseContinuationPoint {
    private _id: string;
    private _browseDirection: number;
    private _referenceTypeId: string;
    private _includeSubtypes: boolean;
    private _nodeClassMask: number;
    private _resultMask: number;
    private _maxReferencesPerNode: number;
    private _referenceOffset: number;
    private _taskCheckListMasks: number;

    constructor(description: UaBrowseDescription, additionalInfo: UaBrowseAdditionalInfo) {
        this._id = description.nodeId.toString();
        this._browseDirection = description.browseDirection;
        this._referenceTypeId = description.referenceTypeId.toString();
        this._includeSubtypes = description.includeSubtypes;
        this._nodeClassMask = description.nodeClassMask;
        this._resultMask = description.resultMask;
        this._maxReferencesPerNode = additionalInfo.maxReferencesPerNode;
        this._referenceOffset = additionalInfo.referenceOffset;
        this._taskCheckListMasks = additionalInfo.taskCheckListMasks;
    }

    get browseDescription(): UaBrowseDescription | null {
        const nodeId = parseUaNodeIdOrNull(this._id);
        const referenceTypeId = parseUaNodeIdOrNull(this._referenceTypeId);

        if (nodeId === null || referenceTypeId === null) return null;

        return new UaBrowseDescription(
            nodeId,
            this._browseDirection,
            referenceTypeId,
            this._includeSubtypes,
            this._nodeClassMask,
            this._resultMask,
        );
    }

    get additionalInfo(): UaBrowseAdditionalInfo {
        return new UaBrowseAdditionalInfo(
            this._maxReferencesPerNode,
            this._referenceOffset,
            this._taskCheckListMasks,
        );
    }

    toByteString(): string {
        const serialized: SerializedBrowseContinuationPoint = {
            id: this._id,
            bDir: this._browseDirection,
            refId: this._referenceTypeId,
            includeSt: this._includeSubtypes,
            ncMask: this._nodeClassMask,
            rltMask: this._resultMask,
            maxRef: this._maxReferencesPerNode,
            offset: this._referenceOffset,
            taskMsk: this._taskCheckListMasks,
        };

        return Buffer.from(JSON.stringify(serialized), "utf-8").toString("base64");
    }

    static fromByteString(continuationPoint: string): UaBrowseContinuationPoint | null {
        if (!continuationPoint) {
            return null;
        }

        try {
            const serialized = JSON.parse(
                Buffer.from(continuationPoint, "base64").toString("utf-8"),
            ) as Partial<SerializedBrowseContinuationPoint>;

            if (!serialized || typeof serialized !== "object") return null;
            if (typeof serialized.bDir !== "number") return null;
            if (typeof serialized.includeSt !== "boolean") return null;
            if (typeof serialized.ncMask !== "number") return null;
            if (typeof serialized.rltMask !== "number") return null;
            if (typeof serialized.maxRef !== "number") return null;
            if (typeof serialized.offset !== "number") return null;
            if (typeof serialized.taskMsk !== "number") return null;

            const nodeId = parseUaNodeIdOrNull(serialized.id);
            const referenceTypeId = parseUaNodeIdOrNull(serialized.refId);

            if (nodeId === null || referenceTypeId === null) {
                return null;
            }

            const browseDescription = new UaBrowseDescription(
                nodeId,
                serialized.bDir,
                referenceTypeId,
                serialized.includeSt,
                serialized.ncMask,
                serialized.rltMask,
            );

            const additionalInfo = new UaBrowseAdditionalInfo(
                serialized.maxRef,
                serialized.offset,
                serialized.taskMsk,
            );

            return new UaBrowseContinuationPoint(browseDescription, additionalInfo);
        } catch {
            return null;
        }
    }
}